docs: creating the sidebar navigation docs (#3306)
twilio-labs/paste: packages/paste-website/src/pages/components/sidebar-navigation/index.mdx

export const meta = {
  title: 'Sidebar Navigation',
  package: '@twilio-paste/sidebar',
  description:

>     'Sidebar navigation comprises a set of user interface controls, designed to facilitate both flat and hierarchical navigation structures within an application.',

Reviewer: ```suggestion
    'Sidebar Navigation controls both flat and hierarchical navigation structures within an application's sidebar.',
```

  SidebarNavigationDisclosureContent,
  SidebarNavigationSeparator,
  useSidebarNavigationDisclosureState,
} from '@twilio-paste/sidebar';
import Changelog from '@twilio-paste/sidebar/CHANGELOG.md';

import {SidebarCategoryRoutes} from '../../../constants';
import packageJson from '@twilio-paste/sidebar/package.json';
import DefaultLayout from '../../../layouts/DefaultLayout';
import {getFeature, getNavigationData} from '../../../utils/api';
import {navigationActions, navigationStateHook} from '../../../component-examples/SidebarNavigationExamples.ts';

export default DefaultLayout;

export const getStaticProps = async () => {
  const navigationData = await getNavigationData();
  const feature = await getFeature('Sidebar');
  return {
    props: {
      data: {
        ...packageJson,
        ...feature,
      },
      navigationData,
    },
  };
};

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/sidebar"
  storybookUrl="?path=/story/components-sidebar-navigation--default"
  description={meta.description}
  data={props.data}
  name="Sidebar Navigation"

>   description="Sidebar navigation comprises a set of user interface controls, designed to facilitate both flat and hierarchical navigation structures within an application."

Reviewer: ```suggestion
  description="Sidebar Navigation controls both flat and hierarchical navigation structures within an application's sidebar."
```

      </SidebarNavigationDisclosure>
      <SidebarNavigationDisclosure>
        <SidebarNavigationDisclosureHeadingWrapper>
          <SidebarNavigationDisclosureHeading icon={<ProductEngageIcon decorative />}>
            Enagage
          </SidebarNavigationDisclosureHeading>
        </SidebarNavigationDisclosureHeadingWrapper>
        <SidebarNavigationDisclosureContent>
          <SidebarNavigationItem href="https://google.com">Overview</SidebarNavigationItem>
        </SidebarNavigationDisclosureContent>
      </SidebarNavigationDisclosure>
      <SidebarNavigationDisclosure>
        <SidebarNavigationDisclosureHeadingWrapper>
          <SidebarNavigationDisclosureHeading icon={<ProductSettingsIcon decorative />}>
            Settings
          </SidebarNavigationDisclosureHeading>
        </SidebarNavigationDisclosureHeadingWrapper>
        <SidebarNavigationDisclosureContent>
          <SidebarNavigationItem href="https://google.com">Overview</SidebarNavigationItem>
        </SidebarNavigationDisclosureContent>
      </SidebarNavigationDisclosure>
    </SidebarNavigation>
  </SidebarBody>
</Box>`}
</LivePreview>

## Guidelines

### About Sidebar Navigation

> Sidebar navigation comprises a set of user interface controls, designed to facilitate both flat and hierarchical navigation structures within an application. The navigation components are presentational and can be utilized to create navigation structures ranging from simple lists to complex, deeply nested structures.
> 
> We recommend limiting the nesting depth to three levels. For navigation beyond this depth, consider also using the [In-page Navigation](/components/in-page-navigation).

Reviewer: ```suggestion
Sidebar Navigation controls both flat and hierarchical navigation structures within an application's sidebar. The navigation components are presentational and can be used to create navigation structures ranging from simple lists to complex, deeply nested structures.

We recommend limiting the nesting depth to 3 levels. For navigation beyond this depth, consider also using the [In-page Navigation](/components/in-page-navigation).
```

### Accessibility

> Nested sidebar navigation employs the [Disclosure pattern](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/).

Reviewer: ```suggestion
Nested sidebar navigation uses the [ARIA Disclosure pattern](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/).
```

- Additional controls or actions should be placed as siblings to each `SidebarNavigationItem` in the navigation, not nested within.
- Every `SidebarNavigationItem` must implement an anchor element by assigning a valid URL to the `href` attribute.
- A `selected` `SidebarNavigationItem` is designated as the "current page" via `aria-current="page"`.

## Examples

### Flat

> In its most basic form, the Sidebar Navigation can be utilized to create a straightforward, flat list of navigation items. This is ideal for smaller applications or single-product use cases.
> 
> Each `SidebarNavigationItem` should be provided with an `href`, a valid URL that the user will navigate to upon interaction. This facilitates opening links in new tabs via cmd/ctrl + clicking or right-clicking on the item, ensuring proper keyboard navigation for keyboard-only users.

Reviewer: ```suggestion
In its most basic form, Sidebar Navigation can be used to create a flat list of navigation items. This is ideal for smaller applications or single-product use cases.

Each `SidebarNavigationItem` should be provided with an `href`, a valid URL that the user will navigate to upon interaction. This facilitates opening links in new tabs via CMD/CTRL + clicking or right-clicking on the item, ensuring proper keyboard navigation for keyboard-only users.
```

      >
        Admin
      </SidebarNavigationItem>
      <SidebarNavigationItem
        href="https://google.com"
        icon={<ProductContactCenterTasksIcon decorative={false} title="Agent dashboard" />}
      >
        Agent dashboard
      </SidebarNavigationItem>
      <SidebarNavigationItem
        href="https://google.com"
        icon={<ProductContactCenterTeamsIcon decorative={false} title="Teams view" />}
      >
        Teams view
      </SidebarNavigationItem>
      <SidebarNavigationItem
        href="https://google.com"
        icon={<ProductContactCenterQueuesIcon decorative={false} title="Queue stats" />}
      >
        Queue stats
      </SidebarNavigationItem>
      <SidebarNavigationItem
        href="https://google.com"
        icon={<ProductPrivacyIcon decorative={false} title="Privacy" />}
      >
        Privacy
      </SidebarNavigationItem>
      <SidebarNavigationItem
        href="https://google.com"
        icon={<ProductUsageIcon decorative={false} title="Insights" />}
      >
        Insights
      </SidebarNavigationItem>
    </SidebarNavigation>
  </SidebarBody>
</Box>`}
</LivePreview>

### Hierarchical

> Sidebar Navigation also supports the creation of hierarchical information architectures using nested disclosures. You can establish "groups" of navigation sections by deploying the `SidebarNavigationSeparator` between groups of items.

Reviewer: ```suggestion
Sidebar Navigation also supports hierarchical information architectures with nested navigation items. You can establish "groups" of navigation sections by using `SidebarNavigationSeparator` between groups of items.
```

Reviewer: I think there's a simpler way to explain this Hierarchical section overall, but I'll just take a look over again when I reread through all the docs.

| `hierarchical?`        | `boolean`         | Set whether you are displating an hirarchical navigation structure.                       |                        |

##### SidebarNavigationItem

| Prop        | Type              | Description                                                                               | Default                     |
| ----------- | ----------------- | ----------------------------------------------------------------------------------------- | --------------------------- |
| `element?`  | `string`          | Overrides the default element name to apply unique styles with the Customization Provider | `'SIDEBAR_NAVIGATION_ITEM'` |
| `children?` | `React.ReactNode` |                                                                                           |                             |
| `href`      | `string`          | String for the url to link to.                                                            |                             |
| `selected`  | `boolean`         | Set to indicate this item matches to the current page the user is on                      |                             |

> | `icon`      | `React.ReactNode` | Any product icons cna be used here                                                        |                             |

Reviewer: ```suggestion
| `icon`      | `React.ReactNode` | Any product icons can be used here                                                        |                             |
```

##### SidebarNavigationDisclosure

| Prop        | Type              | Description                                                                                                             | Default                           |
| ----------- | ----------------- | ----------------------------------------------------------------------------------------------------------------------- | --------------------------------- |
| `element?`  | `string`          | Overrides the default element name to apply unique styles with the Customization Provider                               | `'SIDEBAR_NAVIGATION_DISCLOSURE'` |
| `children?` | `React.ReactNode` |                                                                                                                         |                                   |

> | `baseId?`   | `string`          | Base ID string to base subsequent, dynamic ids on                                                                       |                                   |

Reviewer: ```suggestion
| `baseId?`   | `string`          | Base ID string to base subsequent, dynamic IDs on                                                                       |                                   |
```

| `children?` | `React.ReactNode` |                                                                                           |                                           |

##### SidebarNavigationSeparator

| Prop       | Type     | Description                                                                               | Default                          |
| ---------- | -------- | ----------------------------------------------------------------------------------------- | -------------------------------- |
| `element?` | `string` | Overrides the default element name to apply unique styles with the Customization Provider | `'SIDEBAR_NAVIGATION_SEPARATOR'` |

##### useSidebarNavigationDisclosureState

Pass these as part of an object to the hook.

| Prop       | Type      | Description                                                   | Default |
| ---------- | --------- | ------------------------------------------------------------- | ------- |
| `baseId?`  | `string`  | Set to create a base ID that all dynamic IDs will be based on |         |

> | `visible?` | `boolean` | Set the initial visibilty of the Disclosure                   |         |

Reviewer: ```suggestion
| `visible?` | `boolean` | Set the initial visibility of the Disclosure                   |         |
```

Reviewer: caught one more typo, but I'll approve.

Author: Not much point in approving. When I accept it'll blow it away

Reviewer: it makes me feel better
